pkg/registry/core/pod/strategy.go

```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
+		func(ctr *api.Container, _ podutil.ContainerType) bool {
+			// find the annotation value
+			var (
+				annotation    string
+				hasAnnotation bool
+			)
+			for key, value := range pod.Annotations {
+				if key == api.SeccompContainerAnnotationKeyPrefix+ctr.Name {
+					annotation = value
+					hasAnnotation = true
+					break
+				}
+			}
+
+			// find the field
+			field, hasField := (*api.SeccompProfile)(nil), false
+			if ctr.SecurityContext != nil && ctr.SecurityContext.SeccompProfile != nil {
+				field = ctr.SecurityContext.SeccompProfile
+				hasField = true
+			}
+
+			if hasAnnotation != hasField {
+				// initialize parents if necessary
+				// assign computed values
+				newAnnotation, newField := syncSeccompAnnotationAndField(annotation, field)
+				if newAnnotation != "" {
+					pod.Annotations[v1.SeccompContainerAnnotationKeyPrefix+ctr.Name] = newAnnotation
+				}
+				if newField != nil {
+					if ctr.SecurityContext == nil {
+						ctr.SecurityContext = &api.SecurityContext{}
+					}
+					ctr.SecurityContext.SeccompProfile = newField
+
+				}
+			}
+
+			return true
+		})
+}
+
+// syncSeccompAnnotationAndField takes a pod annotation and a seccomp profile
+// field and returns the synchronized result: The annotation will be converted
+// in an appropriate profile and vice versa.
+func syncSeccompAnnotationAndField(annotation string, field *api.SeccompProfile) (string, *api.SeccompProfile) {
+	// If only seccomp fields are specified, add the corresponding annotations.
+	// This ensures that the fields are enforced even if the node version
+	// trails the API version
+	if field != nil && annotation == "" {
+		switch field.Type {
+		case api.SeccompProfileTypeUnconfined:
+			return v1.SeccompProfileNameUnconfined, nil
+
+		case api.SeccompProfileTypeRuntimeDefault:
+			return v1.SeccompProfileRuntimeDefault, nil
+
+		case api.SeccompProfileTypeLocalhost:
+			return v1.SeccompLocalhostProfileNamePrefix + *field.LocalhostProfile, nil
+		}
+	}
+
+	// If only seccomp annotations are specified, copy the values into the
+	// corresponding fields. This ensures that existing applications continue
+	// to enforce seccomp, and prevents the kubelet from needing to resolve
+	// annotations & fields.
+	if field == nil && annotation != "" {
+		if annotation == v1.SeccompProfileNameUnconfined {
+			return "", &api.SeccompProfile{Type: api.SeccompProfileTypeUnconfined}
+
+		} else if annotation == api.SeccompProfileRuntimeDefault || annotation == api.DeprecatedSeccompProfileDockerDefault {
+			return "", &api.SeccompProfile{Type: api.SeccompProfileTypeRuntimeDefault}
+
+		} else if strings.HasPrefix(annotation, v1.SeccompLocalhostProfileNamePrefix) {
+			localhostProfile := strings.TrimPrefix(annotation, v1.SeccompLocalhostProfileNamePrefix)
+			return "", &api.SeccompProfile{
+				Type:             api.SeccompProfileTypeLocalhost,
+				LocalhostProfile: &localhostProfile,
```

Comment:
this is called pre-validation ... should we handle a zero-length localhostProfile by omitting this?

Reply:
Okay, added a check and unit test for `localhostProfile != ""`.